decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+package openlr.decoder.backtracking;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class BackTrackingGraph<T, L> {
+    private final List<Slice<T, L>> slices;
+    private final Map<String, GraphEdge> edges;
+    private int currentSourceSliceIndex = -1;
+    private GraphEdge<T, L> graphEdge = null;
+    private GraphEdge<T, L> lastSuccessfulEdge = null;
+
+    public BackTrackingGraph(List<Slice<T, L>> slices) {
+        this.slices = slices;
+        this.edges = new HashMap<>();
+    }
+
+    /**
+     * if the user has marked last edge returned by the graph as red;
```

Comment:
No need for `;` at the end of line

Reply:
31890348b9aef835f9f9065ddd8625de1cfab0ea